silkworm/sync/sentry_client.cpp

```diff
@@ -16,137 +16,300 @@
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_id(std::move(message), std::move(peer_public_key)));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_to_random_peers(const OutboundMessage& outbound_message, size_t max_peers) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_to_random_peers(std::move(message), max_peers));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_to_all(const OutboundMessage& outbound_message) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_to_all(std::move(message)));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_by_min_block(const OutboundMessage& outbound_message, BlockNum /*min_block*/, size_t max_peers) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_min_block(std::move(message), max_peers));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+void SentryClient::peer_min_block(const PeerId& peer_id, BlockNum /*min_block*/) {
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    sync_spawn(tasks_, io_context_, sentry_client_->service()->peer_min_block(std::move(peer_public_key)));
+}
 
+void SentryClient::execution_loop() {
     log::set_thread_name("sentry-recv   ");
 
+    auto tasks_completion = [](const std::exception_ptr& ex_ptr) {
+        try {
+            if (ex_ptr) std::rethrow_exception(ex_ptr);
+        } catch (const boost::system::system_error& ex) {
+            if (ex.code() == boost::system::errc::operation_canceled) {
+                log::Debug(kLogTitle) << "execution_loop tasks.wait completed";
+            } else {
+                log::Error(kLogTitle) << "execution_loop tasks.wait system_error: " << ex.what();
+                throw;
+            }
+        } catch (...) {
+            log::Error(kLogTitle) << "execution_loop tasks.wait unexpected exception";
+            throw;
+        }
+    };
+    co_spawn(io_context_, tasks_.wait(), bind_cancellation_slot(tasks_cancellation_signal_.slot(), tasks_completion));
+
     while (!is_stopping()) {
         try {
             connected_ = false;
-            log::Info("SentryClient", {"remote", sentry_addr_}) << " connecting ...";
+            log::Info(kLogTitle) << "connecting ...";
 
             // send current status of the chain
-            hand_shake();
+            handshake();
```

Comment:
IMO we need to remove the re-connection logic (plus handshake & set_status) from here if, as it seems to me, other parts of the application will use the async client.

Reply:
Yes. This will be a part of https://github.com/torquem-ch/silkworm/pull/1008